File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        output = []
        def dfs(cur: List[int]):
            if len(cur) == len(nums):
                output.append(cur)
                return;
            for i, n in enumerate(nums):
                if n in cur: continue
                dfs(cur + [n])

        dfs([])

        return output

File: 46-permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> cur = new ArrayList<>();
        backtrack(nums, cur, result);
        return result;
    }

    private void backtrack(int[] nums, List<Integer> cur, List<List<Integer>> result) {
        if (cur.size() == nums.length) {
            result.add(cur);
            return;
        }
        for (int n : nums) {
            if (cur.contains(n)) continue;
            List<Integer> nextCur = new ArrayList<>();
            nextCur.addAll(cur);
            nextCur.add(n);
            backtrack(nums, nextCur, result);
        }
    }
}